openraft/src/core/mod.rs

```diff
@@ -293,40 +271,39 @@ impl<C: RaftTypeConfig, N: RaftNetworkFactory<C>, S: RaftStorage<C>> RaftCore<C,
         // TODO(xp): this is not necessary.
         self.storage.save_vote(&state.vote).await?;
 
-        self.last_log_id = state.last_log_id;
-        self.vote = state.vote;
-        self.effective_membership = state.effective_membership.clone();
-        self.last_applied = state.last_applied;
+        self.engine = Engine::new(self.id, &state);
+
+        self.engine.state.last_applied = state.last_applied;
 
         // NOTE: The commit index must be determined by a leader after
         // successfully committing a new log to the cluster.
-        self.committed = None;
+        self.engine.state.committed = None;
 
         // Fetch the most recent snapshot in the system.
         if let Some(snapshot) = self.storage.get_current_snapshot().await? {
             self.snapshot_last_log_id = Some(snapshot.meta.last_log_id);
-            self.metrics_flags.set_data_changed();
+            self.engine.metrics_flags.set_data_changed();
         }
 
-        let has_log = if self.last_log_id.is_some() {
+        let has_log = if self.engine.state.last_log_id.is_some() {
```

Comment:
why use String, IMHO a bool is enough, and the same reason with `single`,`is_voter`

Reply:
Previously it was bool, and is a little hard to understand:(

And this is not a hot path, why bother :)

Reply:
when using String, then the code logic will do the same thing with string match, in this way the compiler cannot find problem if miss-type some wrong string to compare. for example, if other day the other auther re-type "has_log" to "has_logs", in this way he has to change two place to make the the logic right.

and if thinking of hard to understand the meaning, the variable name will help, such as `has_log`,etc.

Reply:
Agree. several local variables will help in this case:DDD